cnap/core/metrics.py

```diff
@@ -0,0 +1,182 @@
+"""A Metrics module.
+
+This module provides a class to manage metrics in a Prometheus-based application.
+It provides methods to create, track, and update metrics.
+
+Classes:
+    MetricType: An enum for the type of metric.
+    ErrorMessage: An enum for error messages.
+    MetricsManager: A class to manage metrics in a Prometheus-based application.
+"""
+
+from enum import Enum
+from functools import wraps
+from typing import Any, Callable, Dict, Union
+from threading import Lock
+from prometheus_client import Counter, Gauge, Histogram, Summary
+
+class MetricType(Enum):
+    """An enum for the type of metric.
+
+    This enum defines the types of metrics that can be created.
+    """
+    COUNTER = 'Counter'
+    GAUGE = 'Gauge'
+    HISTOGRAM = 'Histogram'
+    SUMMARY = 'Summary'
+
+
+class ErrorMessage(Enum):
+    """An enum for error messages.
+
+    This enum defines the error messages that can be raised.
+    """
+    METRIC_EXISTS = 'Metric {} already exists'
+    INVALID_METRIC_TYPE = 'Invalid metric type: {}'
+    METRIC_DOES_NOT_EXIST = 'Metric {} does not exist'
+    METRIC_TYPE_MISMATCH = 'Metric {} is not a {}'
+
+
+class MetricsManager:
+    """A class to manage metrics in a Prometheus-based application.
+
+    This class provides methods to create, track, and update metrics.
+
+    Attributes:
+        metrics (dict): A dictionary of all metrics.
+    """
+
+    _METRIC_TYPE_MAP = {
+        MetricType.COUNTER: Counter,
+        MetricType.GAUGE: Gauge,
+        MetricType.HISTOGRAM: Histogram,
+        MetricType.SUMMARY: Summary,
+    }
+
+    def __init__(self):
+        """Initializes the MetricsManager with an empty metrics dictionary.
+
+        This method initializes the MetricsManager with an empty metrics dictionary.
+        """
+        self.metrics: Dict[str, Union[Counter, Gauge, Histogram, Summary]] = {}
+        self.lock = Lock()
+
+    def get_metric(self, metric_name: str, metric_type: type) -> Union[
+        Counter, Gauge, Histogram, Summary]:
+        """Returns the metric with the specified name and type.
+
+        Args:
+            metric_name (str): The name of the metric.
+            metric_type (type): The type of the metric.
+
+        Returns:
+            Union[Counter, Gauge, Histogram, Summary]: The metric with the specified name and type.
+
+        Raises:
+            ValueError: If the specified metric does not exist or the metric type is invalid.
+        """
+        with self.lock:
+            metric = self.metrics.get(metric_name)
+            if metric is None:
+                raise ValueError(ErrorMessage.METRIC_DOES_NOT_EXIST.value.format(metric_name))
+
+            if not isinstance(metric, metric_type):
```

Comment:
`metric_type` doen't seem to be a suitable argument for the `isinstance` function.
TypeError: isinstance() arg 2 must be a type or tuple of types

Reply:
fixed.